File: trans/bots.py
"""Bot per TRANS.

Tre livelli:

- ``facile``   mosse legali a caso, scommessa approssimata;
- ``normale``  scommessa stimata con playout Monte Carlo, gioco euristico;
- ``esperto``  piu' simulazioni e conteggio delle carte gia' uscite.
"""

from __future__ import annotations

import random

from .cards import Card, Suit, full_deck, trick_key
from .engine import Game, Phase
from .rules import RoundKind

LEVELS = ("facile", "normale", "esperto")

_SAMPLES = {"facile": 0, "normale": 40, "esperto": 140}


# --------------------------------------------------------------------- utils


def _best_in_trick(trick: list[tuple[int, Card]], trump: Suit | None) -> tuple[int, Card] | None:
    if not trick:
        return None
    lead = trick[0][1].suit
    return max(trick, key=lambda pc: trick_key(pc[1], lead, trump))


def _currently_wins(card: Card, trick: list[tuple[int, Card]], trump: Suit | None) -> bool:
    """True se la carta batte tutto quello che c'e' gia' sul tavolo."""
    best = _best_in_trick(trick, trump)
    if best is None:
        return True  # apre lui la presa
    lead = trick[0][1].suit
    return card.beats(best[1], lead, trump)


def _strength(card: Card, lead: Suit | None, trump: Suit | None) -> tuple[int, int]:
    return trick_key(card, lead or card.suit, trump)


# ------------------------------------------------------------------ scommessa


def _greedy_playout(
    hands: list[list[Card]], leader: int, trump: Suit | None, rng: random.Random
) -> list[int]:
    """Gioca una mano intera con una politica rozza ma simmetrica.

    Ogni giocatore prende la presa se puo' farlo a poco prezzo, altrimenti
    scarta la carta piu' bassa. Serve solo a stimare quante prese vale una mano.
    """
    n = len(hands)
    tricks = [0] * n
    hands = [list(h) for h in hands]
    while hands[leader]:
        trick: list[tuple[int, Card]] = []
        for step in range(n):
            p = (leader + step) % n
            hand = hands[p]
            lead = trick[0][1].suit if trick else None
            legal = [c for c in hand if c.suit is lead] if lead else list(hand)
            if not legal:
                legal = list(hand)
            winners = [c for c in legal if _currently_wins(c, trick, trump)]
            if winners:
                card = min(winners, key=lambda c: _strength(c, lead, trump))
            else:
                card = min(legal, key=lambda c: _strength(c, lead, trump))
            hand.remove(card)
            trick.append((p, card))
        best = _best_in_trick(trick, trump)
        assert best is not None
        leader = best[0]
        tricks[leader] += 1
    return tricks


def estimate_tricks(
    hand: list[Card],
    n_players: int,
    trump: Suit | None,
    unseen: list[Card],
    samples: int,
    rng: random.Random,
) -> float:
    """Numero medio di prese che questa mano porta a casa, via Monte Carlo."""
    if not hand or samples <= 0:
        return len(hand) / max(n_players, 1)
    cards_each = len(hand)
    total = 0
    for _ in range(samples):
        pool = list(unseen)
        rng.shuffle(pool)
        hands = [list(hand)]
        ok = True
        for i in range(n_players - 1):
            chunk = pool[i * cards_each : (i + 1) * cards_each]
            if len(chunk) < cards_each:
                ok = False
                break
            hands.append(chunk)
        if not ok:
            return len(hand) / max(n_players, 1)
        total += _greedy_playout(hands, leader=0, trump=trump, rng=rng)[0]
    return total / samples


def choose_bid(game: Game, player_index: int, rng: random.Random | None = None) -> int:
    rng = rng or random.Random()
    player = game.players[player_index]
    spec = game.spec
    level = player.bot_level if player.bot_level in LEVELS else "normale"

    if spec.blind_bidding:
        # Al buio non si vede niente: si punta sulla quota equa, con un filo di varianza.
        fair = spec.cards / game.n
        guess = int(round(fair + rng.uniform(-0.6, 0.6)))
        return max(0, min(spec.cards, guess))

    if level == "facile":
        options = game.legal_bids(player_index)
        naive = max(0, min(spec.cards, int(round(spec.cards / game.n))))
        pool = [b for b in options if abs(b - naive) <= 1] or options
        return rng.choice(pool)

    unseen = [c for c in full_deck() if c not in player.hand]
    expected = estimate_tricks(
        player.hand, game.n, game.trump, unseen, _SAMPLES[level], rng
    )
    bid = int(round(expected))
    options = game.legal_bids(player_index)
    if bid not in options:
        bid = min(options, key=lambda b: (abs(b - expected), b))
    return bid


# ---------------------------------------------------------------------- gioco


def _wants_trick(game: Game, player_index: int) -> bool:
    """Se al bot conviene prendere questa presa."""
    spec = game.spec
    player = game.players[player_index]
    if spec.kind is RoundKind.MISERE:
        return False
    if player.bid is None:
        return False
    needed = player.bid - player.tricks
    remaining = len(player.hand)
    if needed <= 0:
        # Scommessa gia' raggiunta: prendere ancora la fa saltare.
        # Se e' gia' saltata, ogni presa in piu' vale comunque 1 punto.
        return needed < 0
    if needed > remaining:
        # Impossibile centrarla: tanto vale raccogliere punti da 1.
        return True
    return True


def _unseen_cards(game: Game, player_index: int) -> list[Card]:
    known = set(game.players[player_index].hand) | set(game.played_cards)
    known |= {c for _, c in game.current_trick}
    return [c for c in full_deck() if c not in known]


def choose_card(game: Game, player_index: int, rng: random.Random | None = None) -> Card:
    rng = rng or random.Random()
    legal = game.legal_cards(player_index)
    if not legal:
        raise RuntimeError("nessuna carta giocabile")
    if len(legal) == 1:
        return legal[0]

    player = game.players[player_index]
    level = player.bot_level if player.bot_level in LEVELS else "normale"
    if level == "facile":
        return rng.choice(legal)

    trump = game.trump
    trick = game.current_trick
    lead = game.lead_suit
    want = _wants_trick(game, player_index)
    last_to_play = len(trick) == game.n - 1

    if not trick:
        return _lead_card(game, player_index, legal, want, level, rng)

    winners = [c for c in legal if _currently_wins(c, trick, trump)]
    losers = [c for c in legal if c not in winners]

    if want:
        if not winners:
            return min(legal, key=lambda c: _strength(c, lead, trump))
        if last_to_play:
            return min(winners, key=lambda c: _strength(c, lead, trump))
        # Restano giocatori dopo di noi: serve una carta che regga davvero.
        safe = [c for c in winners if _is_safe(c, game, player_index, level)]
        pool = safe or winners
        return min(pool, key=lambda c: _strength(c, lead, trump))

    # Vogliamo perdere la presa: scarichiamo la carta piu' alta che non prende.
    if losers:
        return max(losers, key=lambda c: _strength(c, lead, trump))
    return max(legal, key=lambda c: _strength(c, lead, trump))


def _is_safe(card: Card, game: Game, player_index: int, level: str) -> bool:
    """La carta batte tutto quello che gli avversari possono ancora avere?"""
    if level != "esperto":
        return card.rank >= 12 or (game.trump is not None and card.suit is game.trump)
    trump = game.trump
    lead = game.lead_suit or card.suit
    threshold = trick_key(card, lead, trump)
    for other in _unseen_cards(game, player_index):
        if trick_key(other, lead, trump) > threshold:
            return False
    return True


def _lead_card(
    game: Game, player_index: int, legal: list[Card], want: bool, level: str, rng: random.Random
) -> Card:
    """Scelta della carta di uscita."""
    trump = game.trump
    if want:
        if level == "esperto":
            sure = [c for c in legal if _is_safe(c, game, player_index, level)]
            if sure:
                return max(sure, key=lambda c: _strength(c, c.suit, trump))
        non_trump_high = [c for c in legal if trump is None or c.suit is not trump]
        if non_trump_high and max(c.rank for c in non_trump_high) >= 13:
            return max(non_trump_high, key=lambda c: c.rank)
        return max(legal, key=lambda c: _strength(c, c.suit, trump))

    # Uscire perdendo: seme lungo e carta bassa, evitando la briscola.
    side = [c for c in legal if trump is None or c.suit is not trump]
    pool = side or legal
    return min(pool, key=lambda c: (c.rank, _strength(c, c.suit, trump)))


# ------------------------------------------------------------------- driver


def act(game: Game, player_index: int, rng: random.Random | None = None) -> dict:
    """Fa muovere il bot di turno e descrive la mossa (per il log della UI)."""
    player = game.players[player_index]
    if game.phase is Phase.BIDDING:
        bid = choose_bid(game, player_index, rng)
        game.place_bid(player.id, bid)
        return {"action": "bid", "value": bid}
    if game.phase is Phase.PLAYING:
        card = choose_card(game, player_index, rng)
        game.play_card(player.id, card)
        return {"action": "play", "card": card.code}
    raise RuntimeError(f"il bot non puo' agire in fase {game.phase}")

File: site/js/bots.js
/* Bot di TRANS in JavaScript — porto di trans/bots.py.
 *
 * Stessi tre livelli e stesse euristiche della versione Python: la
 * dichiarazione si stima con playout Monte Carlo sulle mani possibili, il
 * gioco decide se conviene prendere e con quale carta.
 */

import {
  Phase,
  RoundKind,
  fullDeck,
  shuffle,
  rankOf,
  suitOf,
  trickKey,
} from "./engine.js";

export const LEVELS = ["facile", "normale", "esperto"];

const SAMPLES = { facile: 0, normale: 40, esperto: 140 };

/* ------------------------------------------------------------------ utili */

function bestInTrick(trick, trump) {
  if (!trick.length) return null;
  const lead = suitOf(trick[0].card);
  let best = trick[0];
  for (const play of trick.slice(1)) {
    if (trickKey(play.card, lead, trump) > trickKey(best.card, lead, trump)) best = play;
  }
  return best;
}

/** La carta batte tutto quello che c'e' gia' sul tavolo? */
function currentlyWins(card, trick, trump) {
  const best = bestInTrick(trick, trump);
  if (!best) return true; // apre lui la presa
  const lead = suitOf(trick[0].card);
  return trickKey(card, lead, trump) > trickKey(best.card, lead, trump);
}

function strength(card, lead, trump) {
  return trickKey(card, lead || suitOf(card), trump);
}

function minBy(items, score) {
  return items.reduce((a, b) => (score(b) < score(a) ? b : a));
}

function maxBy(items, score) {
  return items.reduce((a, b) => (score(b) > score(a) ? b : a));
}

/* ------------------------------------------------------------ dichiarazione */

/**
 * Gioca una mano intera con una politica rozza ma uguale per tutti: si prende
 * la presa se costa poco, altrimenti si scarta la carta piu' bassa. Serve solo
 * a stimare quanto vale una mano, non a giocare bene.
 */
function greedyPlayout(hands, leader, trump) {
  const n = hands.length;
  const tricks = new Array(n).fill(0);
  const remaining = hands.map((h) => [...h]);

  while (remaining[leader].length) {
    const trick = [];
    for (let step = 0; step < n; step++) {
      const p = (leader + step) % n;
      const hand = remaining[p];
      const lead = trick.length ? suitOf(trick[0].card) : null;
      let legal = lead ? hand.filter((c) => suitOf(c) === lead) : [...hand];
      if (!legal.length) legal = [...hand];
      const winners = legal.filter((c) => currentlyWins(c, trick, trump));
      const card = minBy(winners.length ? winners : legal, (c) => strength(c, lead, trump));
      hand.splice(hand.indexOf(card), 1);
      trick.push({ player: p, card });
    }
    leader = bestInTrick(trick, trump).player;
    tricks[leader] += 1;
  }
  return tricks;
}

/** Prese medie che questa mano porta a casa, via Monte Carlo. */
export function estimateTricks(hand, nPlayers, trump, unseen, samples, random = Math.random) {
  if (!hand.length || samples <= 0) return hand.length / Math.max(nPlayers, 1);
  const cardsEach = hand.length;
  if (unseen.length < cardsEach * (nPlayers - 1)) return hand.length / Math.max(nPlayers, 1);

  let total = 0;
  for (let s = 0; s < samples; s++) {
    const pool = shuffle(unseen, random);
    const hands = [[...hand]];
    for (let i = 0; i < nPlayers - 1; i++) {
      hands.push(pool.slice(i * cardsEach, (i + 1) * cardsEach));
    }
    total += greedyPlayout(hands, 0, trump)[0];
  }
  return total / samples;
}

/** La dichiarazione legale piu' vicina a quella che il bot vorrebbe fare. */
function nearestLegal(game, playerIndex, wanted) {
  const options = game.legalBids(playerIndex);
  if (!options.length) throw new Error("nessuna dichiarazione possibile");
  // Come la tupla (abs(b - wanted), b) del Python: distanza, poi il piu' basso.
  return options.reduce((best, b) => {
    const da = Math.abs(best - wanted);
    const db = Math.abs(b - wanted);
    return db < da || (db === da && b < best) ? b : best;
  });
}

export function chooseBid(game, playerIndex, random = Math.random) {
  const player = game.players[playerIndex];
  const spec = game.spec;
  const level = LEVELS.includes(player.botLevel) ? player.botLevel : "normale";

  if (spec.blindBidding) {
    // Al buio non si vede niente: quota equa, con un filo di varianza.
    const fair = spec.cards / game.n;
    return nearestLegal(game, playerIndex, fair + (random() * 1.2 - 0.6));
  }

  if (level === "facile") {
    const options = game.legalBids(playerIndex);
    const naive = spec.cards / game.n;
    const pool = options.filter((b) => Math.abs(b - naive) <= 1);
    const from = pool.length ? pool : options;
    return from[Math.floor(random() * from.length)];
  }

  const inHand = new Set(player.hand);
  const unseen = fullDeck().filter((c) => !inHand.has(c));
  const expected = estimateTricks(player.hand, game.n, game.trump, unseen, SAMPLES[level], random);
  return nearestLegal(game, playerIndex, expected);
}

/* ------------------------------------------------------------------ gioco */

/**
 * A perdere, prendere tutte le prese vale +15 invece di -5 a presa.
 *
 * Ci si prova solo a colpo sicuro: non averne ancora persa una e avere in mano
 * solo carte imbattibili. Senza, il bot a sei prese su sette giocherebbe per
 * perdere l'ultima, portandosi a casa -30 invece di +15.
 */
function moonIsOn(game, playerIndex) {
  const player = game.players[playerIndex];
  if (!player.hand.length) return false;
  const played = game.spec.cards - player.hand.length;
  if (player.tricks !== played) return false; // una presa gia' persa
  const level = LEVELS.includes(player.botLevel) ? player.botLevel : "normale";
  return player.hand.every((card) => isSafe(card, game, playerIndex, level));
}

/** Al bot conviene prendere questa presa? */
function wantsTrick(game, playerIndex) {
  const spec = game.spec;
  const player = game.players[playerIndex];
  if (spec.kind === RoundKind.MISERE) return moonIsOn(game, playerIndex);
  if (player.bid === null) return false;

  const needed = player.bid - player.tricks;
  // Dichiarazione gia' raggiunta: prendere ancora la fa saltare. Se e' gia'
  // saltata, ogni presa in piu' vale comunque 1 punto.
  if (needed <= 0) return needed < 0;
  return true;
}

function unseenCards(game, playerIndex) {
  const known = new Set([
    ...game.players[playerIndex].hand,
    ...game.playedCards,
    ...game.currentTrick.map((p) => p.card),
  ]);
  return fullDeck().filter((c) => !known.has(c));
}

/** La carta batte tutto quello che gli avversari possono ancora avere? */
function isSafe(card, game, playerIndex, level) {
  const trump = game.trump;
  if (level !== "esperto") {
    return rankOf(card) >= 12 || (trump !== null && suitOf(card) === trump);
  }
  const lead = game.leadSuit || suitOf(card);
  const threshold = trickKey(card, lead, trump);
  return !unseenCards(game, playerIndex).some((other) => trickKey(other, lead, trump) > threshold);
}

function leadCard(game, playerIndex, legal, want, level) {
  const trump = game.trump;
  if (want) {
    if (level === "esperto") {
      const sure = legal.filter((c) => isSafe(c, game, playerIndex, level));
      if (sure.length) return maxBy(sure, (c) => strength(c, suitOf(c), trump));
    }
    const side = legal.filter((c) => !trump || suitOf(c) !== trump);
    if (side.length && Math.max(...side.map(rankOf)) >= 13) {
      return maxBy(side, rankOf);
    }
    return maxBy(legal, (c) => strength(c, suitOf(c), trump));
  }
  // Uscire perdendo: seme lungo e carta bassa, evitando la briscola.
  const side = legal.filter((c) => !trump || suitOf(c) !== trump);
  const pool = side.length ? side : legal;
  return minBy(pool, (c) => rankOf(c) * 1000 + strength(c, suitOf(c), trump));
}

export function chooseCard(game, playerIndex, random = Math.random) {
  const legal = game.legalCards(playerIndex);
  if (!legal.length) throw new Error("nessuna carta giocabile");
  if (legal.length === 1) return legal[0];

  const player = game.players[playerIndex];
  const level = LEVELS.includes(player.botLevel) ? player.botLevel : "normale";
  if (level === "facile") return legal[Math.floor(random() * legal.length)];

  const trump = game.trump;
  const trick = game.currentTrick;
  const lead = game.leadSuit;
  const want = wantsTrick(game, playerIndex);
  const lastToPlay = trick.length === game.n - 1;

  if (!trick.length) return leadCard(game, playerIndex, legal, want, level);

  const winners = legal.filter((c) => currentlyWins(c, trick, trump));
  const losers = legal.filter((c) => !winners.includes(c));

  if (want) {
    if (!winners.length) return minBy(legal, (c) => strength(c, lead, trump));
    if (lastToPlay) return minBy(winners, (c) => strength(c, lead, trump));
    // Restano giocatori dopo di noi: serve una carta che regga davvero.
    const safe = winners.filter((c) => isSafe(c, game, playerIndex, level));
    return minBy(safe.length ? safe : winners, (c) => strength(c, lead, trump));
  }

  // Vogliamo perdere la presa: scarichiamo la carta piu' alta che non prende.
  if (losers.length) return maxBy(losers, (c) => strength(c, lead, trump));
  return maxBy(legal, (c) => strength(c, lead, trump));
}

/** Fa muovere il bot di turno. */
export function act(game, playerIndex, random = Math.random) {
  const player = game.players[playerIndex];
  if (game.phase === Phase.BIDDING) {
    const bid = chooseBid(game, playerIndex, random);
    game.placeBid(player.id, bid);
    return { action: "bid", value: bid };
  }
  if (game.phase === Phase.PLAYING) {
    const card = chooseCard(game, playerIndex, random);
    game.playCard(player.id, card);
    return { action: "play", card };
  }
  throw new Error(`il bot non puo' agire in fase ${game.phase}`);
}
